### ISSUE TITLE
RET504 false positive via decorator

### ISSUE THREAD

**Issue by charliermarsh** [charliermarsh/ruff#3394]
Here's an interesting case where variable is assigned to, referenced, and then returned:
https://github.com/KyleKing/pytest_cache_assert/blob/47d0b538528a9c9cdbc182d2b276462a2aa0fd4d/tests/test_flask.py#L20
```python
def _create_app() -> Flask:
    app = Flask(__name__)

    @app.route('/hello')
    def hello() -> str:
        """Hello endpoint."""
        return 'Hello, World!'

    return app  # noqa: RET504
```
Seems like something that shouldn't be flagged.

Perhaps instead of an assignment count, heuristic could be reference count. In other words, excluding return statement, flag if the only reference was assignment.

_Originally posted by @gdub in https://github.com/charliermarsh/ruff/issues/2950#issuecomment-1456740839_
